src/main/java/com/linkedin/kmf/services/KeyedProduceService.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright 2016 LinkedIn Corp. Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.linkedin.kmf.services;
+
+import com.linkedin.kmf.common.Utils;
+import com.linkedin.kmf.producer.BaseProducerRecord;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.kafka.common.utils.Utils.murmur2;
+
+
+public class KeyedProduceService extends ProduceService implements Service {
+  private static final Logger LOG = LoggerFactory.getLogger(KeyedProduceService.class);
+  private static final String METRIC_GROUP_NAME = "keyed-produce-service";
+  private final ConcurrentMap<String, AtomicLong> _nextIndexPerKey;
+
+  public KeyedProduceService(Map<String, Object> props, String name) throws Exception {
+    super(props, name);
+    _nextIndexPerKey = new ConcurrentHashMap<>();
+
+  }
+
+  @Override
+  public void start() {
+    if (_running.compareAndSet(false, true)) {
+      initializeStateForPartitions();
+      _handleNewPartitionsExecutor.scheduleWithFixedDelay(new NewPartitionHandler(), 40_000, 40_000, TimeUnit.MILLISECONDS);
+      LOG.info(_name + "/ProduceService started");
+    }
+  }
+
+  private void initializeStateForPartitions() {
+    Map<Integer, String> keyMapping = generateKeyMappings();
+    int partitionNum = _partitionNum.get();
+    for (int partition = 0; partition < partitionNum; partition++) {
+      String key = keyMapping.get(partition);
+      //This is what preserves sequence numbers across restarts
+      if (!_nextIndexPerKey.containsKey(key)) {
+        _nextIndexPerKey.put(key, new AtomicLong(0));
+        _sensors.addPartitionSensors(partition);
+      }
+      _produceExecutor.scheduleWithFixedDelay(new ProduceRunnable(partition, key), _produceDelayMs, _produceDelayMs, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private Map<Integer, String> generateKeyMappings() {
+    int partitionNum = _partitionNum.get();
+    HashMap<Integer, String> keyMapping = new HashMap<>();
+
+    int keyIndex = 0;
+    while (keyMapping.size() < partitionNum) {
```

Comment:
We chatted about this a while ago. The issue is that mirror maker does not propagate partition numbers across clusters so we have to use keys. The partitions will then be assigned by a modulus of a hash of the key. If we don't purposefully select our keys, we could have a non-uniform assignment across partitions.